### ISSUE TITLE
NestedIfDepth: handles only nested if and ignores nesting in else blocks

### ISSUE THREAD

**Issue by rdiachenko** [checkstyle/checkstyle#13493]
```bash
/var/tmp/checkstyle% cat TestClass.java 
public class TestClass {
    void fun() {
        // nesting == 2
        if (true) {
        } else {
            if (true) {
            } else {
                if (false) { // I would expect violation here, but I get nothing
                } else {
                }
            }
        }

        // nesting == 2
        if (true) {
            if (true) {
                if (false) { // violation here occurred as expected
                } else {
                }
            } else {
            }
        } else {
        }
    }
}

/var/tmp/checkstyle% cat TestConfig.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
      <module name="NestedIfDepth"/>
    </module>
</module>

/var/tmp/checkstyle% java -jar checkstyle-10.12.2-all.jar -c TestConfig.xml TestClass.java
Starting audit...
[ERROR] /private/var/tmp/checkstyle/TestClass.java:17:17: Nested if-else depth is 2 (max allowed is 1). [NestedIfDepth]
Audit done.
Checkstyle ends with 1 errors.
```

I would expect the following 2 violations:
```
[ERROR] /private/var/tmp/checkstyle/TestClass.java:8:17: Nested if-else depth is 2 (max allowed is 1). [NestedIfDepth]
[ERROR] /private/var/tmp/checkstyle/TestClass.java:17:17: Nested if-else depth is 2 (max allowed is 1). [NestedIfDepth]
```
But there is only one violation.

**Comment by romani** [comment]
I am ok to approve.
@rnveach , what do you think ?

**Comment by rnveach** [comment]
I was under the impression `NestedIf` was to follow similar behavior like `NestedFor` and `NestedTry` where we are only checking the depth of those specific tokens and ignoring all others. It is named `NestedIf` and not `NestedIfElse`. `NestedTry` does examine depth in catch/finally blocks.

This is not connected to the Google config. According to a30cd217f341e61b12e017b59a8a129283583d61 where the checks were added (no issue), it always only looked at only if statements.

@romani You will have to confirm if this is a bug and we always intended to verify the sibling tokens , or this is not an issue and/or should be a new check.

**Comment by romani** [comment]
In documentation of javadoc and xdoc it mentioned if-else
https://github.com/checkstyle/checkstyle/commit/a30cd217f341e61b12e017b59a8a129283583d61#diff-faa722f9db184c13b7dcafb6daaf163369597ec1db41ad0d59103283d52a3abeR26

So it means might be simply mistake of implementation.

NestedFor does not have second scope block, so it is not comparable.

So we need to decide to extend NestedIf and NestedTry to control nesting in other sibling blocks.
NestedIf - to control nesting of IF token under IF token (else branches should be counted). https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_IF

NestedTry - to control nesting of TRY under. https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_TRY , https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_CATCH

I see all sibling blocks under IF and TRY,  so I vote to count them.

Looks like https://checkstyle.org/checks/misc/descendanttoken.html#DescendantToken was introduced before such specific Checks. People always cared about amount of blocks in method, to prevent too much nesting. 

I think it will be good for NestedIf and NestedTry to restrict more.

**Comment by romani** [comment]
@rdiachenko , @nrmancuso , please vote also.

**Comment by rdiachenko** [comment]
Makes sense to me to extend existing logic for `NestedTry` and `NestedIf` to check related sibling blocks. That is what I initially expected from the `NestedIf` at least.

> NestedTry - to control nesting of TRY under. 
> https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_TRY , 
> https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_CATCH

We may need to include `finally` branch as well https://checkstyle.sourceforge.io/apidocs/com/puppycrawl/tools/checkstyle/api/TokenTypes.html#LITERAL_FINALLY

**Comment by romani** [comment]
Yes, `finally` should also. All blocks under try/if.

**Comment by nrmancuso** [comment]
It is debatable whether we should add support for this, it will probably cause users a lot of pain to resolve all new violations. Should we consider creating new checks, i.e. `NestedIfElse`? If nothing else, we should create tests in the scope of this issue with XPathSuppression module that replicates current behavior of this check (suppressing all new violations), and share the module here in the issue.

**Comment by romani** [comment]
In my mind it just a defect in check, and yes there will be more violations as any time we fix false negatives, release and user update checkstyle version.

I agree to provide users xpath suppression to ignore violations under ELSE token.

**Comment by harshit4311** [comment]
@romani @rdiachenko I would love working on this issue! Could you please assign it to me?
